src/main/java/rx/internal/operators/OperatorMerge.java

```diff
@@ -71,715 +72,878 @@
+        /** What was its index in the innerSubscribers array? Accessed if emitting == true. */
+        int lastIndex;
+        
+        /** An empty array to avoid creating new empty arrays in removeInner. */ 
+        static final InnerSubscriber<?>[] EMPTY = new InnerSubscriber<?>[0];
 
-        private void handleScalarSynchronousObservable(ScalarSynchronousObservable<? extends T> t) {
-            // fast-path for scalar, synchronous values such as Observable.from(int)
-            /**
-             * Without this optimization:
-             * 
-             * <pre> {@code
-             * Benchmark                                          (size)   Mode   Samples        Score  Score error    Units
-             * r.o.OperatorMergePerf.oneStreamOfNthatMergesIn1         1  thrpt         5  2,418,452.409   130572.665    ops/s
-             * r.o.OperatorMergePerf.oneStreamOfNthatMergesIn1      1000  thrpt         5     5,690.456       94.958    ops/s
-             * r.o.OperatorMergePerf.merge1SyncStreamOfN         1000000  thrpt         5          takes too long
-             * 
-             * With this optimization:
-             * 
-             * r.o.OperatorMergePerf.merge1SyncStreamOfN               1  thrpt         5  5,475,300.198   156741.334    ops/s
-             * r.o.OperatorMergePerf.merge1SyncStreamOfN            1000  thrpt         5    68,932.278     1311.023    ops/s
-             * r.o.OperatorMergePerf.merge1SyncStreamOfN         1000000  thrpt         5       64.405        0.611    ops/s
-             * } </pre>
-             * 
-             */
-            if (mergeProducer.requested == Long.MAX_VALUE) {
-                handleScalarSynchronousObservableWithoutRequestLimits(t);
-            } else {
-                handleScalarSynchronousObservableWithRequestLimits(t);
+        /** In fast mode, this counts the number of active subscriptions, including the one to upstream. */
+        volatile int fastWip;
+        @SuppressWarnings("rawtypes")
+        static final AtomicIntegerFieldUpdater<MergeSubscriber> FAST_WIP
+        = AtomicIntegerFieldUpdater.newUpdater(MergeSubscriber.class, "fastWip");
+        
+        volatile boolean fastMode;
+        
+        /** Guarded by this. */
+        Object[] fastArray;
+        /** Guarded by this. */
+        int fastSize;
+        
+        /** Fast, unbounded emission report contents of 'errors'. */
+        static final Object REPORT_ERROR = new Object();
+        /** Fast, unbounded emission complete. */
+        static final Object REPORT_COMPLETED = new Object();
+        
+        public MergeSubscriber(Subscriber<? super T> child, boolean delayErrors, int maxConcurrent) {
+            this.child = child;
+            this.delayErrors = delayErrors;
+            this.maxConcurrent = maxConcurrent;
+            this.nl = NotificationLite.instance();
+            this.innerGuard = new Object();
+            this.innerSubscribers = EMPTY;
+            this.request(0);
```

Comment:
Curious as to why you are calling this here instead of using `onStart`?

I think this is safe to do from within the constructor due to when this all gets initialized, but on first read it felt wrong for this to happen here. Probably not an issue but would like to understand your thinking for not following the normal pattern of using `onStart`.